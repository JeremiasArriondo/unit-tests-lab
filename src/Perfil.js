import React, { Component } from 'react';

// Componente Perfil.
class Perfil extends Component {

    // Constructor para el manejo de props y state.
    constructor(props) {
        super(props);
        this.state = {
            usuario: this.props.usuario
        }
    }

    // Muestra en pantalla cada una de las cartas 
    // en componente renderizado
    render() {
        const {usuario} = this.state;
        return (
            <div className="celda">
                <p><span className="titulo">Nombre:</span><span className="nombre"> {usuario.nombre}</span></p>
                <p><span className="titulo">E-mail:</span><span className="email"> {usuario.email}</span></p>
                <p><span className="titulo">Teléfono:</span><span className="telefono"> {usuario.telefono}</span></p>
                <p><span className="titulo">Código Postal:</span><span className="codigoPostal"> {usuario.codigoPostal}</span></p>
            </div>
        );
    }
}

export default Perfil;