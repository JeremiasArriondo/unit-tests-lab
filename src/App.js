import React, { Component } from 'react';
import axios from 'axios';
import Perfil from './Perfil';

// Componente App.
class App extends Component {
  
  // Estado inicial para state.
  state = {
    usuarios: [],
    isCargandoPagina: true,
    error: null,
    color: '#ccffff'
  }

  // Evento onChange sobre div contenedor,
  // el cual hace que cambie el color del background
  onChange = () => {
    if (this.state.color === '#ccffff') {
      this.setState({ color: '#ffff99' });
    } else {
      this.setState({ color: '#ccffff' });
    }
  }

  // Invocación al API endpoint de typicode
  // y mapeo de resultados
  obtenerUsuarios() {
    axios
      .get("https://jsonplaceholder.typicode.com/users")
      .then(response =>
        response.data.map(usuario => ({
          nombre: `${usuario.name}`,
          nombreUsuario: `${usuario.username}`,
          email: `${usuario.email}`,
          telefono: `${usuario.phone}`,
          codigoPostal: `${usuario.address.zipcode}`
        }))
      )
      .then(usuarios => {
        this.setState({
          usuarios,
          isCargandoPagina: false
        });
      })
      .catch(error => this.setState({ error, isCargandoPagina: false }));
  }

  // Una vez que el componente se inserta al árbol React
  // Se invoca al API endpoint y se mapean los usuarios
  componentDidMount() {
    this.obtenerUsuarios();
  }

  // Se pasan los resultados de la invocación
  // del API endpoint al componente Perfil
  // via iteración
  render() {
    const { isCargandoPagina, usuarios } = this.state;
    return (
      <div style={{ backgroundColor: this.state.color }} onClick={this.onChange} className="container">
        <h2>Listado de Usuarios</h2>
        <div className="row">
          {!isCargandoPagina ? (
            usuarios.map(usuario => {
              return (
                <Perfil usuario={usuario} />
              );
            })
          ) : (
              <p>Cargando página...</p>
            )}
        </div>
      </div>
    );
  }
}

export default App;
