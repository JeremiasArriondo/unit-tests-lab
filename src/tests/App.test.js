import React from 'react';
import { shallow } from "enzyme";
import App from '../App';
import Perfil from '../Perfil';


describe('Test suite Cartas obtener usuarios', () => {
    let wrapper;
    /*
    Hago uso de la funcion beforeEach para evitar la repeticion en las pruebas
    */
    beforeEach(() => {
        wrapper = shallow (
            <App />
        );
    });
    /*
    Punto a, aqui busco matchear o hacer coincidir con el snapshot, verificando que exista un archivo con
    el mismo nombre que el de donde se encuentra el test con extension .snap dentro de la carpeta __snapshots__
    */
    test('Debe coincidir con Snapshot', () => {
        expect(wrapper).toMatchSnapshot();
    });
    /*
    Punto b, simplemente busco el titulo que posee una etiqueta h2 en el componente App, utilizo toEqual para comparar
    el argumento con lo que posee el titulo.
    */
    test('El titulo debe ser "Listado de Usuarios"', () => {
        expect(
            wrapper.find("h2").text()
        ).toEqual("Listado de Usuarios");
    });
    /*
    Punto c, hago uso de la funcion containsMatchingElement para evaluar el texto de cargando página, esta funcion 
    retorna un booleano cuando coincide con un elemento del arbol de renderizado
    */
    test('Debe existir un loader con texto "Cargando página"', () => {
        expect(
            wrapper.containsMatchingElement(
                <p>Cargando página...</p>
            )
        ).toBe(true);
    });
    
    describe('Test suite componente Perfil', () => {
        let wrapper;
        
        beforeEach(() => {
            //Creo un mock de un usuario para utilizar en uno de los tests
            const user = {
                nombre: 'Jeremias',
                email: 'unit-test@mock.org.us',
                telefono: 202109,
                codigoPostal: 2705
            };
            wrapper = shallow (
                <Perfil usuario={user}/>
            );

        });
        /*
        Punto d, aqui busco matchear o hacer coincidir con el snapshot, verificando que exista un archivo con
        el mismo nombre que el de donde se encuentra el test con extension .snap dentro de la carpeta __snapshots__
        */
        test('Perfil debe coincidir con Snapshot', () => {
            expect(wrapper).toMatchSnapshot();
        });
        /*
        Punto e, utilizo el metodo .state() que me devuelve el estado del nodo de envoltura 
        para hacer uso del nombre de accesorio y devolver ese valor, luego utilizo toEqual para comparar
        */
        test('El usuario debe tener una dirección de e-mail unit-test@mock.org.us', () => {
            expect(
                wrapper.state().usuario.email
            ).toEqual("unit-test@mock.org.us");
        });
        /*
        Punto f. En este punto hago uso de find, el cual lo utilizó para encontrar el nodo que necesito, luego utilizo lenght 
        para comparar la cantidad de elementos con toBe
        */
        test('Cada carta debe tener 4 elementos de párrafo con clase "titulo"', () => {
            expect(
                wrapper.find("p").find(".titulo").length
            ).toBe(4);
        });
        /*
        Punto g. El metodo some devuelve un booleano si algun selector coincide con el proporcionado, lo utilice en
        conjunto con toBe para hacer coincidir el test con un valor booleano.
        */
        test('El elemento div dentro de cada carta debe tener una clase "celda"', () => {
            expect(
                wrapper.find('div').some('.celda')
            ).toBe(true);
        });
    });
    /*
    Punto h. En este punto vuelvo a utilizar el estado, en este caso, el estado inicial, buscando
    en su propiedad color, luego con toEqual lo comparo
    */
    test('El color de fondo debe ser inicialmente "#ccffff"', () => {
        expect(
            wrapper.state().color
        ).toEqual("#ccffff");
    });
    /*
    Punto i. En este punto con find busco al div contenedor, simulo el click y busco
    si la propiedad del color cambia al valor con el cual lo comparo.
    */
    test('El color de fondo debe ser "#ffff99" al hacer click en el div contenedor', () => {
        wrapper.find('div.container').simulate('click');
        expect(
            wrapper.state().color
        ).toEqual('#ffff99');
    });
})
