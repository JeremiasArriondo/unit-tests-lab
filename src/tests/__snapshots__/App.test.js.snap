// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test suite Cartas obtener usuarios Debe coincidir con Snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <App />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <h2>
          Listado de Usuarios
        </h2>,
        <div
          className="row"
        >
          <p>
            Cargando página...
          </p>
        </div>,
      ],
      "className": "container",
      "onClick": [Function],
      "style": Object {
        "backgroundColor": "#ccffff",
      },
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Listado de Usuarios",
        },
        "ref": null,
        "rendered": "Listado de Usuarios",
        "type": "h2",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <p>
            Cargando página...
          </p>,
          "className": "row",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Cargando página...",
          },
          "ref": null,
          "rendered": "Cargando página...",
          "type": "p",
        },
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h2>
            Listado de Usuarios
          </h2>,
          <div
            className="row"
          >
            <p>
              Cargando página...
            </p>
          </div>,
        ],
        "className": "container",
        "onClick": [Function],
        "style": Object {
          "backgroundColor": "#ccffff",
        },
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Listado de Usuarios",
          },
          "ref": null,
          "rendered": "Listado de Usuarios",
          "type": "h2",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <p>
              Cargando página...
            </p>,
            "className": "row",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Cargando página...",
            },
            "ref": null,
            "rendered": "Cargando página...",
            "type": "p",
          },
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    Symbol(enzyme.__providerValues__): undefined,
  },
  Symbol(enzyme.__providerValues__): Map {},
  Symbol(enzyme.__childContext__): null,
}
`;

exports[`Test suite Cartas obtener usuarios Test suite componente Perfil Perfil debe coincidir con Snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Perfil
    usuario={
      Object {
        "codigoPostal": 2705,
        "email": "unit-test@mock.org.us",
        "nombre": "Jeremias",
        "telefono": 202109,
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <p>
          <span
            className="titulo"
          >
            Nombre:
          </span>
          <span
            className="nombre"
          >
             
            Jeremias
          </span>
        </p>,
        <p>
          <span
            className="titulo"
          >
            E-mail:
          </span>
          <span
            className="email"
          >
             
            unit-test@mock.org.us
          </span>
        </p>,
        <p>
          <span
            className="titulo"
          >
            Teléfono:
          </span>
          <span
            className="telefono"
          >
             
            202109
          </span>
        </p>,
        <p>
          <span
            className="titulo"
          >
            Código Postal:
          </span>
          <span
            className="codigoPostal"
          >
             
            2705
          </span>
        </p>,
      ],
      "className": "celda",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <span
              className="titulo"
            >
              Nombre:
            </span>,
            <span
              className="nombre"
            >
               
              Jeremias
            </span>,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Nombre:",
              "className": "titulo",
            },
            "ref": null,
            "rendered": "Nombre:",
            "type": "span",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                " ",
                "Jeremias",
              ],
              "className": "nombre",
            },
            "ref": null,
            "rendered": Array [
              " ",
              "Jeremias",
            ],
            "type": "span",
          },
        ],
        "type": "p",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <span
              className="titulo"
            >
              E-mail:
            </span>,
            <span
              className="email"
            >
               
              unit-test@mock.org.us
            </span>,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "E-mail:",
              "className": "titulo",
            },
            "ref": null,
            "rendered": "E-mail:",
            "type": "span",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                " ",
                "unit-test@mock.org.us",
              ],
              "className": "email",
            },
            "ref": null,
            "rendered": Array [
              " ",
              "unit-test@mock.org.us",
            ],
            "type": "span",
          },
        ],
        "type": "p",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <span
              className="titulo"
            >
              Teléfono:
            </span>,
            <span
              className="telefono"
            >
               
              202109
            </span>,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Teléfono:",
              "className": "titulo",
            },
            "ref": null,
            "rendered": "Teléfono:",
            "type": "span",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                " ",
                202109,
              ],
              "className": "telefono",
            },
            "ref": null,
            "rendered": Array [
              " ",
              202109,
            ],
            "type": "span",
          },
        ],
        "type": "p",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <span
              className="titulo"
            >
              Código Postal:
            </span>,
            <span
              className="codigoPostal"
            >
               
              2705
            </span>,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Código Postal:",
              "className": "titulo",
            },
            "ref": null,
            "rendered": "Código Postal:",
            "type": "span",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                " ",
                2705,
              ],
              "className": "codigoPostal",
            },
            "ref": null,
            "rendered": Array [
              " ",
              2705,
            ],
            "type": "span",
          },
        ],
        "type": "p",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <p>
            <span
              className="titulo"
            >
              Nombre:
            </span>
            <span
              className="nombre"
            >
               
              Jeremias
            </span>
          </p>,
          <p>
            <span
              className="titulo"
            >
              E-mail:
            </span>
            <span
              className="email"
            >
               
              unit-test@mock.org.us
            </span>
          </p>,
          <p>
            <span
              className="titulo"
            >
              Teléfono:
            </span>
            <span
              className="telefono"
            >
               
              202109
            </span>
          </p>,
          <p>
            <span
              className="titulo"
            >
              Código Postal:
            </span>
            <span
              className="codigoPostal"
            >
               
              2705
            </span>
          </p>,
        ],
        "className": "celda",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <span
                className="titulo"
              >
                Nombre:
              </span>,
              <span
                className="nombre"
              >
                 
                Jeremias
              </span>,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Nombre:",
                "className": "titulo",
              },
              "ref": null,
              "rendered": "Nombre:",
              "type": "span",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  " ",
                  "Jeremias",
                ],
                "className": "nombre",
              },
              "ref": null,
              "rendered": Array [
                " ",
                "Jeremias",
              ],
              "type": "span",
            },
          ],
          "type": "p",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <span
                className="titulo"
              >
                E-mail:
              </span>,
              <span
                className="email"
              >
                 
                unit-test@mock.org.us
              </span>,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "E-mail:",
                "className": "titulo",
              },
              "ref": null,
              "rendered": "E-mail:",
              "type": "span",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  " ",
                  "unit-test@mock.org.us",
                ],
                "className": "email",
              },
              "ref": null,
              "rendered": Array [
                " ",
                "unit-test@mock.org.us",
              ],
              "type": "span",
            },
          ],
          "type": "p",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <span
                className="titulo"
              >
                Teléfono:
              </span>,
              <span
                className="telefono"
              >
                 
                202109
              </span>,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Teléfono:",
                "className": "titulo",
              },
              "ref": null,
              "rendered": "Teléfono:",
              "type": "span",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  " ",
                  202109,
                ],
                "className": "telefono",
              },
              "ref": null,
              "rendered": Array [
                " ",
                202109,
              ],
              "type": "span",
            },
          ],
          "type": "p",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <span
                className="titulo"
              >
                Código Postal:
              </span>,
              <span
                className="codigoPostal"
              >
                 
                2705
              </span>,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Código Postal:",
                "className": "titulo",
              },
              "ref": null,
              "rendered": "Código Postal:",
              "type": "span",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  " ",
                  2705,
                ],
                "className": "codigoPostal",
              },
              "ref": null,
              "rendered": Array [
                " ",
                2705,
              ],
              "type": "span",
            },
          ],
          "type": "p",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    Symbol(enzyme.__providerValues__): undefined,
  },
  Symbol(enzyme.__providerValues__): Map {},
  Symbol(enzyme.__childContext__): null,
}
`;
